aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/listeners/SegmentListener.java

```diff
@@ -33,18 +45,40 @@ default void beforeEndSegment(Segment segment) {
     }
 
     /**
-     * afterEndSegment is invoked after a segment is ended by the recorder and immediately before it is emitted to the daemon.
+     * afterEndSegment is invoked after a segment is ended by the recorder and emitted to the daemon.
      * The segment must not be manipulated. Attempts to do so will raise an exception.
      *
      * @param segment
      * The segment that has just ended
      */
-    default void afterEndSegment(Segment segment) {
+    default void afterEndSegment(final Segment segment) {
+
+    }
+
+    /**
+     * beforeEndSubsegment is invoked just before a subsegment is ended by the recorder.
+     * The subsegment can be manipulated, e.g. with putAnnotation.
```

Comment:
`putAnnotation` doesn't seem to be a good example because even with `final` you can still call `subsegment.putAnnotation`. BTW I'm confused by "The subsegment cannot be manipulated". The subsegment itself is surely mutable. The argument is a referrence of the subsegment. The method can never replace the original subsegment, so what is the point to allow manipulating the referrence?  (ref https://stackoverflow.com/questions/1249917/final-variable-manipulation-in-java)

Reply:
This is a fair callout. The (sub)segment can't be modified in `afterEndSegment` because it is invoked after the segment is emitted to the daemon, and attempting to modify it would raise an `AlreadyEmittedException`. I'll update the docs to clarify this.

I added the `final` keyword to indicate its immutability, but you're right that making a parameter `final` does not make it immutable. I'll remove the `final` to avoid confusion.